base/applications/mspaint/mouse.cpp

```diff
@@ -58,216 +61,444 @@ BOOL nearlyEqualPoints(INT x0, INT y0, INT x1, INT y1)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+};
+
+// TOOL_ZOOM
+struct ZoomTool : ToolBase
+{
+    ZoomTool() : ToolBase(TOOL_ZOOM)
+    {
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        if (bLeftButton)
+        {
+            if (toolsModel.GetZoom() < MAX_ZOOM)
+                zoomTo(toolsModel.GetZoom() * 2, x, y);
+        }
+        else
+        {
+            if (toolsModel.GetZoom() > MIN_ZOOM)
+                zoomTo(toolsModel.GetZoom() / 2, x, y);
+        }
+    }
+};
+
+// TOOL_PEN
+struct PenTool : ToolBase
+{
+    PenTool() : ToolBase(TOOL_PEN)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+        {
+            Line(m_hdc, last.x, last.y, x, y, m_fg, 1);
+            SetPixel(m_hdc, x, y, m_fg);
+        }
+        else
+        {
+            Line(m_hdc, last.x, last.y, x, y, m_bg, 1);
+            SetPixel(m_hdc, x, y, m_bg);
+        }
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        if (bLeftButton)
+            SetPixel(m_hdc, x, y, m_fg);
+        else
+            SetPixel(m_hdc, x, y, m_bg);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_BRUSH
+struct BrushTool : ToolBase
+{
+    BrushTool() : ToolBase(TOOL_BRUSH)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Brush(m_hdc, last.x, last.y, x, y, m_fg, toolsModel.GetBrushStyle());
+        else
+            Brush(m_hdc, last.x, last.y, x, y, m_bg, toolsModel.GetBrushStyle());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
```

Comment:
Note: usually for drawing tools (like pen, brush..) there's no real action done on mouse-up.